homeassistant/helpers/template.py

```diff
@@ -914,14 +914,21 @@ def device_entities(hass: HomeAssistant, _device_id: str) -> Iterable[str]:
     return [entry.entity_id for entry in entries]
 
 
-def device_id(hass: HomeAssistant, entity_id: str) -> str | None:
-    """Get a device ID from an entity ID."""
-    if not isinstance(entity_id, str) or "." not in entity_id:
-        raise TemplateError(f"Must provide an entity ID, got {entity_id}")  # type: ignore
+def device_id(hass: HomeAssistant, entity_id_or_device_name: str) -> str | None:
+    """Get a device ID from an entity ID or device name."""
     entity_reg = entity_registry.async_get(hass)
-    entity = entity_reg.async_get(entity_id)
+    entity = entity_reg.async_get(entity_id_or_device_name)
     if entity is None:
```

Comment:
Maybe invert this check and return the entity device id if true. Then we can outdent below and have more of the code outdented.